### DISCUSSION TITLE
Nuxt3 on Plesk with Phusion Passenger

### DISCUSSION THREAD

**Discussion by fabianwohlfart** [nuxt/nuxt#16106] [General]
I just wanted to share this little hint to get Nuxt3 deployed with Phusion Passenger.
As PR can not load the `./.output/server/index.mjs` directly you can simply add an `./entry.js` file with...

```javascript
import("./.output/server/index.mjs");
```

...to your folder and use this as your application file.

__Update 19. Okt. 22:__
Early Hints are introduced as experimental in rc12.
This is breaking. Compare https://github.com/nuxt/framework/issues/8306

Solution is to disable this feature (for now)

```
// nuxt.config.ts
export default defineNuxtConfig({
  ...
  experimental: {
    writeEarlyHints: false,
  },
  ...
})
```

**Comment by hoomb** [comment]
thank you Sir. That was life saving.

**Reply by Tofandel** [reply]
As part of the build script you can do:

`"nuxt build && echo 'import(\"./index.mjs\");' > .output/server/index.cjs && mkdir -p .output/tmp && touch .output/tmp/restart.txt"`

Then `rsync -ar --no-perms .output/ your-ssh-server:/your/deploy/path` and set `App root` to `/your/deploy/path`, `Document root` to `/your/deploy/path/public` and startup file to `server/index.cjs`

**Comment by subjacked** [comment]
@fabianwohlfart Thank  You! You saved me hours!

**Comment by Pawel2505** [comment]
@fabianwohlfart could you please describe it a bit more specific? I mean what directories are 'Application Root' and 'Document Root'? It looks that you moved the entry.js file in the parent directory for .output directory and Plesk does not allow placing 'Application Startup File' anywhere else than in the Application Root. I am struggling with it for hours :( Any advice?

**Reply by fabianwohlfart** [reply]
@Pawel2505 
Document Root would be the /public folder.
Application Root is your main root folder where app.vue, nuxt.config.js etc. are.
From the plesk docs:

> When we’re talking about web apps (e.g. Node.js based apps or similar), the “Document root” is the location where the static assets can be found. Meanwhile, the “Application root” is the root directory where the app itself is located. In most cases, the “Document root” is a directory inside the “Application root” directory (e.g. “public”). 

Next to the application root files you put the entry.js file.
You choose the entry.js file as the Application File.

Caveat: I tested it with a pre-RC release of nuxt, but i guess it works also with the current RC.

**Reply by Tofandel** [reply]
You should only be deploying the content of the .output folder, nothing else, it's a self contained application and doesn't require node modules (the needed ones are copied in .output/server/node_modules) and your dev env, vue files and stuff (it would be a security issue to leave them there with application root pointing to them), as such I would set the index.cjs in the .output/server folder as well

Then set the application root to the `.output`, the document root to `.output/public` and the startup file to `server/index.cjs`

**Comment by Eliophot** [comment]
Hello, thanks for this.
I have another issue after that whith Phusion Passenger : 
```
(node:2795) UnhandledPromiseRejectionWarning: SyntaxError: Unexpected token '?'
    at Loader.moduleStrategy (internal/modules/esm/translators.js:140:18)
(node:2795) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
(node:2795) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```
Any ideas ?

**Reply by fabianwohlfart** [reply]
That's inside your Javascript, nothing to do with PP.

**Comment by shaylash** [comment]
Thanks bro !!!!

**Comment by jaguar19961** [comment]
I spend 2 days to search why the application did not work... Thanks!!!!

**Comment by negope** [comment]
thanks so much amigo, i  really appreciate it :)

**Comment by vanling** [comment]
Can anyone else confirm this is now broken on Plesk with rc12?

**Reply by fabianwohlfart** [reply]
I think it's because Early Hints got introduced.

https://github.com/nuxt/framework/discussions/8296

**Reply by vanling** [reply]
Yes, also breaks my default nginx deployments without Plesk :( 

https://github.com/nuxt/framework/issues/8306, you could work around this by disabling this early hints feature. (yet to test this on Plesk)

**Reply by fabianwohlfart** [reply]
Ok, I copy this here for now.
So people do not have to search.

```
export default defineNuxtConfig({
  ...
  experimental: {
    writeEarlyHints: false,
  },
  ...
})
```

**Comment by mathiasayivor** [comment]
THIS IS A LIFESAVER!!!
I basically just created `entry.js` in the project root (where nuxt.config is located) and included `import("./.output/server/index.mjs");`, then used `entry.js` as the Application Startup File.

Perhaps this instruction should be included in the deployment docs?

**Reply by vanling** [reply]
Would also be nice if this entry file could be created with a preset

```
export default {
  nitro: {
    preset: 'plesk'
  }
}
```

**Reply by TheAlexLichter** [reply]
PR welcome for https://github.com/unjs/nitro ☺️

**Comment by mohamad68** [comment]
@fabianwohlfart  
hi can you write help for deploying nuxt 3 on Plesk?
 I can not deploy nuxt 3 on Plesk please help

**Comment by JoJuhasz** [comment]
Thanks, very usefull :)

**Comment by vanling** [comment]
This topic is still saving my ass.
I'm currently using this small script to upload  Nuxt3 to plesk ftp with `npm run deploy'

It creates the `tmp/restart.txt` that will restart nodejs app on first reload of page and the entry.js file.

https://gist.github.com/vanling/021d77a0544362dff5338cc7bef31882

_package.json_
```
"ftp-deploy": "2.4.3",

"deploy": "nuxt build && node scripts/upload-ftp.js",
```

<img width="618" alt="image" src="https://user-images.githubusercontent.com/50699657/218308472-006c76d9-e056-406a-9615-e917433c7e1c.png">

**Comment by asolopovas** [comment]
Is there a way to run entry with `"type": "module"` in package json? 
```
/opt/cpanel/ea-ruby27/root/usr/share/passenger/helper-scripts/node-loader.js:80
	return originalRequire.apply(this, arguments);
                        ^

Error [ERR_REQUIRE_ESM]: require() of ES Module /home/user/test/app.js from /opt/cpanel/ea-ruby27/root/usr/share/passenger/helper-scripts/node-loader.js not supported.
app.js is treated as an ES module file as it is a .js file whose nearest parent package.json contains "type": "module" which declares all .js files in that package scope as ES modules.
Instead either rename app.js to end in .cjs, change the requiring code to use dynamic import() which is available in all CommonJS modules, or change "type": "module" to "type": "commonjs" in /home/user/test/package.json to treat all .js files as CommonJS (using .mjs for all ES modules instead).

    at Module.require (/opt/cpanel/ea-ruby27/root/usr/share/passenger/helper-scripts/node-loader.js:80:25)
    at loadApplication (/opt/cpanel/ea-ruby27/root/usr/share/passenger/helper-scripts/node-loader.js:243:2)
    at setupEnvironment (/opt/cpanel/ea-ruby27/root/usr/share/passenger/helper-scripts/node-loader.js:214:2)
    at Object.<anonymous> (/opt/cpanel/ea-ruby27/root/usr/share/passenger/helper-scripts/node-loader.js:133:1) {
  code: 'ERR_REQUIRE_ESM'
}

Node.js v20.10.0
```

**Reply by fabianwohlfart** [reply]
Did you try the suggested solutions?

**Reply by asolopovas** [reply]
If I remove `type=module` it works but with it throws an error. I am not sure if this breaks something that's why I asked the question

**Reply by fabianwohlfart** [reply]
I would remove "type": "module" and work around this. What kind of errors do you get then?
